File: scripts/sku_current_risk_diagnostic.py
#!/usr/bin/env python3
"""Build a read-only, component-level SKU Current Risk diagnostic.

This deliberately does not produce a signal, priority, forecast, reorder action,
or production threshold. It only reshapes existing SKU facts into auditable
observations for design review.
"""

from __future__ import annotations

import argparse
import json
from pathlib import Path
from typing import Any


def number(value: Any) -> float | None:
    if value is None or value == "":
        return None
    try:
        result = float(value)
    except (TypeError, ValueError):
        return None
    return result if result == result else None


def rounded(value: float | None, digits: int = 4) -> float | None:
    return None if value is None else round(value, digits)


def cover_band(cover: float | None) -> str:
    if cover is None:
        return "UNKNOWN"
    if cover <= 0:
        return "ZERO"
    if cover < 2:
        return "UNDER_2_WEEKS"
    if cover <= 4:
        return "2_TO_4_WEEKS"
    return "OVER_4_WEEKS"


def inbound_completion_band(completion: float | None) -> str:
    if completion is None:
        return "UNKNOWN"
    if completion <= 0:
        return "NO_INBOUND"
    if completion < 1:
        return "PARTIAL"
    return "COMPLETE_OR_OVER"


def diagnostic_row(style: dict[str, Any], sku: dict[str, Any]) -> dict[str, Any]:
    order = number(sku.get("orderQty"))
    inbound = number(sku.get("inboundQty"))
    stock = number(sku.get("erpStockQty"))
    velocity = number(sku.get("weighted4CompletedWeekQty"))
    selling_age_velocity = number(sku.get("sellingAgeVelocityQtyPerWeek"))
    completed_history = sku.get("completedWeeklyHistory") or []
    completed_weeks = len(completed_history)
    completed_sales = sum(number(item.get("qty")) or 0 for item in completed_history)
    inbound_completion = inbound / order if order and order > 0 else None
    remaining_order = order - inbound if order is not None and inbound is not None else None
    stock_cover = number(sku.get("stockCoverWeeks"))
    selling_age_stock_cover = number(sku.get("sellingAgeStockCoverWeeks"))

    data_quality = []
    if completed_weeks == 0 or velocity is None or velocity <= 0:
        data_quality.append("NO_COMPLETED_VELOCITY")
    if order is None:
        data_quality.append("ORDER_MISSING")
    if inbound is None:
        data_quality.append("INBOUND_MISSING")
    if stock is None:
        data_quality.append("ERP_STOCK_MISSING")
    if stock_cover is None:
        data_quality.append("STOCK_COVER_MISSING")

    return {
        "sku": sku.get("sku"),
        "styleCode": sku.get("styleCode") or style.get("styleCode"),
        "colorCode": sku.get("colorCode"),
        "category": sku.get("category") or style.get("category"),
        "productGroup": sku.get("productGroup") or style.get("productGroup"),
        "season": sku.get("season") or style.get("season"),
        "facts": {
            "completedWeeks": completed_weeks,
            "completedSalesQty": rounded(completed_sales, 4),
            "completedVelocityQtyPerWeek": rounded(velocity, 4),
            "legacyVelocityQtyPerWeek": rounded(velocity, 4),
            "sellingAgeVelocityQtyPerWeek": rounded(selling_age_velocity, 4),
            "lastCompleteWeekQty": number(sku.get("lastCompleteWeekQty")),
            "previousCompleteWeekQty": number(sku.get("previousCompleteWeekQty")),
            "completedWeekWowPct": number(sku.get("completedWeekWow")),
            "erpStockQty": stock,
            "orderQty": order,
            "inboundQty": inbound,
            "inboundCompletionPct": rounded(inbound_completion * 100 if inbound_completion is not None else None),
            "remainingOrderQty": rounded(remaining_order),
            "salesTrend": sku.get("salesTrend"),
            "stockCoverWeeks": stock_cover,
            "legacyStockCoverWeeks": stock_cover,
            "sellingAgeStockCoverWeeks": selling_age_stock_cover,
        },
        "observations": {
            "stockCoverBand": cover_band(stock_cover),
            "inboundCompletionBand": inbound_completion_band(inbound_completion),
            "hasRemainingOrder": remaining_order is not None and remaining_order > 0,
            "hasErpStock": stock is not None and stock > 0,
            "trend": sku.get("salesTrend") or "UNKNOWN",
        },
        "dataQuality": data_quality,
    }


def build_diagnostic(payload: dict[str, Any]) -> dict[str, Any]:
    rows = []
    for style in (payload.get("styles") or {}).values():
        for sku in style.get("skus") or []:
            rows.append(diagnostic_row(style, sku))
    rows.sort(key=lambda row: str(row.get("sku") or ""))
    return {
        "schemaVersion": "sku-current-risk-diagnostic-v1",
        "diagnosticOnly": True,
        "source": payload.get("meta", {}),
        "definition": {
            "velocity": "existing weighted4CompletedWeekQty; completed weeks only",
            "sellingAgeVelocity": "separate sellingAgeVelocityQtyPerWeek field; completed weeks only; leading pre-launch zero weeks excluded; WTD excluded",
            "inboundCompletion": "inboundQty / orderQty * 100 when orderQty > 0",
            "remainingOrder": "max is not applied; raw orderQty - inboundQty",
            "trend": "existing salesTrend classification",
            "stockCover": "existing stockCoverWeeks",
            "sellingAgeStockCover": "separate sellingAgeStockCoverWeeks field; display-only comparison",
            "bands": "descriptive observation bands only; not business thresholds",
        },
        "summary": {
            "skuCount": len(rows),
            "noCompletedVelocityCount": sum("NO_COMPLETED_VELOCITY" in row["dataQuality"] for row in rows),
            "sellingAgeVelocityCount": sum(row["facts"]["sellingAgeVelocityQtyPerWeek"] is not None for row in rows),
            "sellingAgeStockCoverCount": sum(row["facts"]["sellingAgeStockCoverWeeks"] is not None for row in rows),
            "stockCoverBandCounts": counts(rows, "stockCoverBand"),
            "inboundCompletionBandCounts": counts(rows, "inboundCompletionBand"),
            "trendCounts": counts(rows, "trend"),
        },
        "rows": rows,
    }


def counts(rows: list[dict[str, Any]], key: str) -> dict[str, int]:
    result: dict[str, int] = {}
    for row in rows:
        value = row["observations"].get(key, "UNKNOWN")
        result[value] = result.get(value, 0) + 1
    return dict(sorted(result.items()))


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--input", type=Path, default=Path("data/sku-latest.json"))
    parser.add_argument("--output", type=Path, default=Path("data/sku-current-risk-diagnostic.json"))
    args = parser.parse_args()
    payload = json.loads(args.input.read_text(encoding="utf-8"))
    result = build_diagnostic(payload)
    args.output.parent.mkdir(parents=True, exist_ok=True)
    args.output.write_text(json.dumps(result, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")
    print(json.dumps(result["summary"], ensure_ascii=False))


if __name__ == "__main__":
    main()

File: public/js/sku-current-risk-diagnostic.js
(() => {
  const state = { rows: [], filtered: [], sort: "sku", dir: 1 };
  const $ = (id) => document.getElementById(id);
  const labels = {
    UNDER_2_WEEKS: "2주 미만",
    "2_TO_4_WEEKS": "2~4주",
    OVER_4_WEEKS: "4주 초과",
    ZERO: "0주",
    UNKNOWN: "미확인",
    NO_INBOUND: "입고 없음",
    PARTIAL: "부분 입고",
    COMPLETE_OR_OVER: "완료/초과",
  };
  const nf = (v, digits = 0) => v == null || Number.isNaN(Number(v)) ? "-" : Number(v).toLocaleString("ko-KR", { maximumFractionDigits: digits });
  const pct = (v) => v == null ? "-" : `${nf(v, 1)}%`;
  const ageCover = (cover) => cover == null ? "미확인" : `${nf(cover, 1)}주`;

  function value(row, key) {
    const f = row.facts;
    if (key === "velocity") return f.completedVelocityQtyPerWeek;
    if (key === "sellingAgeVelocity") return f.sellingAgeVelocityQtyPerWeek;
    if (key === "stock") return f.erpStockQty;
    if (key === "inbound") return f.inboundCompletionPct;
    if (key === "remaining") return f.remainingOrderQty;
    if (key === "coverWeeks") return f.stockCoverWeeks;
    return row[key];
  }

  function renderMetrics(summary) {
    const cards = [
      ["SKU rows", summary.skuCount, "SKU-first universe"],
      ["Legacy velocity missing", summary.noCompletedVelocityCount, "WTD excluded"],
      ["Selling-age velocity", summary.sellingAgeVelocityCount || 0, "separate field only"],
      ["Selling-age cover", summary.sellingAgeStockCoverCount || 0, "display comparison"],
    ];
    $("metrics").innerHTML = cards.map(([a, b, c]) => `<div class="metric"><small>${a}</small><strong>${nf(b)}</strong><span>${c}</span></div>`).join("");
  }

  function options(id, values, map = (x) => x) {
    const el = $(id);
    [...new Set(values.filter(Boolean))].sort().forEach((v) => {
      const o = document.createElement("option");
      o.value = v;
      o.textContent = map(v);
      el.append(o);
    });
  }

  function render() {
    const q = $("search").value.trim().toUpperCase();
    const filters = { group: $("group").value, category: $("category").value, trend: $("trend").value, cover: $("cover").value, inbound: $("inbound").value };
    state.filtered = state.rows.filter((r) => {
      const hay = `${r.sku} ${r.styleCode} ${r.colorCode}`.toUpperCase();
      return (!q || hay.includes(q))
        && (filters.group === "all" || r.productGroup === filters.group)
        && (filters.category === "all" || r.category === filters.category)
        && (filters.trend === "all" || r.observations.trend === filters.trend)
        && (filters.cover === "all" || r.observations.stockCoverBand === filters.cover)
        && (filters.inbound === "all" || r.observations.inboundCompletionBand === filters.inbound);
    });
    const sorted = [...state.filtered].sort((a, b) => {
      const av = value(a, state.sort), bv = value(b, state.sort);
      if (typeof av === "string" || typeof bv === "string") return String(av || "").localeCompare(String(bv || "")) * state.dir;
      return ((Number(av) || -Infinity) - (Number(bv) || -Infinity)) * state.dir;
    });
    $("status").textContent = `${nf(sorted.length)}개 표시 / ${nf(state.rows.length)}개 전체 · 행을 누르면 원천 관찰값을 확인합니다`;
    $("rows").innerHTML = sorted.length ? sorted.map(rowHtml).join("") : `<tr><td colspan="11" class="empty">조건에 맞는 SKU가 없습니다.</td></tr>`;
    document.querySelectorAll("#rows tr[data-sku]").forEach((tr) => tr.addEventListener("click", () => openDrawer(state.rows.find((r) => r.sku === tr.dataset.sku))));
  }

  function rowHtml(r) {
    const f = r.facts;
    const q = r.dataQuality.length ? `<span class="tag orange">${r.dataQuality.includes("NO_COMPLETED_VELOCITY") ? "완료 속도 없음" : "확인 필요"}</span>` : `<span class="tag">정상 수집</span>`;
    return `<tr data-sku="${r.sku}">
      <td><b>${r.sku}</b><br><span class="muted">${r.styleCode || "-"} · ${r.colorCode || "-"}</span></td>
      <td>${r.productGroup || "-"}</td>
      <td>${r.category || "-"}</td>
      <td class="num">${nf(f.completedVelocityQtyPerWeek, 1)} <span class="muted">/주</span></td>
      <td class="num">${nf(f.sellingAgeVelocityQtyPerWeek, 1)} <span class="muted">/주</span></td>
      <td class="num">${nf(f.erpStockQty)}</td>
      <td class="num">${pct(f.inboundCompletionPct)}</td>
      <td class="num">${nf(f.remainingOrderQty)}</td>
      <td><span class="tag blue">${r.observations.trend}</span></td>
      <td>${labels[r.observations.stockCoverBand] || r.observations.stockCoverBand}<br><span class="muted">legacy ${nf(f.stockCoverWeeks, 1)}주 · age ${ageCover(f.sellingAgeStockCoverWeeks)}</span></td>
      <td>${q}</td>
    </tr>`;
  }

  function openDrawer(r) {
    if (!r) return;
    const f = r.facts;
    $("drawerTitle").textContent = r.sku;
    $("drawerSub").textContent = `${r.styleCode || "-"} · ${r.colorCode || "-"} · ${r.productGroup || "-"} / ${r.category || "-"}`;
    const facts = [
      ["완료 판매 주차", `${nf(f.completedWeeks)}주`],
      ["완료 판매 수량", nf(f.completedSalesQty)],
      ["Legacy 완료 속도", `${nf(f.completedVelocityQtyPerWeek, 1)} /주`],
      ["Selling-age 속도", `${nf(f.sellingAgeVelocityQtyPerWeek, 1)} /주`],
      ["최근 / 직전 주", `${nf(f.lastCompleteWeekQty)} / ${nf(f.previousCompleteWeekQty)}`],
      ["ERP 재고", nf(f.erpStockQty)],
      ["발주 / 입고", `${nf(f.orderQty)} / ${nf(f.inboundQty)}`],
      ["입고 완료", pct(f.inboundCompletionPct)],
      ["잔여 발주", nf(f.remainingOrderQty)],
      ["판매 추이", f.salesTrend || "미확인"],
      ["Legacy Stock Cover", f.stockCoverWeeks == null ? "미확인" : `${nf(f.stockCoverWeeks, 1)}주`],
      ["Selling-age Cover", f.sellingAgeStockCoverWeeks == null ? "미확인" : `${nf(f.sellingAgeStockCoverWeeks, 1)}주`],
    ];
    $("drawerFacts").innerHTML = facts.map(([a, b]) => `<div class="fact"><small>${a}</small><b>${b}</b></div>`).join("");
    $("drawerQuality").innerHTML = r.dataQuality.length
      ? `<div class="quality">데이터 상태: ${r.dataQuality.join(" · ")}<br><span style="font-weight:600">완료되지 않은 주차나 WTD를 추정하여 보정하지 않았습니다.</span></div>`
      : `<div class="quality" style="background:#eef7f1;color:#41635a">데이터 상태: 추가 플래그 없음</div>`;
    document.body.classList.add("open");
  }

  ["search", "group", "category", "trend", "cover", "inbound"].forEach((id) => $(id).addEventListener("input", render));
  document.querySelectorAll("th button").forEach((b) => b.addEventListener("click", () => {
    const key = b.dataset.sort;
    state.dir = state.sort === key ? state.dir * -1 : 1;
    state.sort = key;
    render();
  }));
  $("close").addEventListener("click", () => document.body.classList.remove("open"));
  $("drawerBg").addEventListener("click", () => document.body.classList.remove("open"));

  fetch("data/sku-current-risk-diagnostic.json", { cache: "no-store" })
    .then((r) => r.json())
    .then((data) => {
      state.rows = data.rows || [];
      renderMetrics(data.summary || {});
      options("category", state.rows.map((r) => r.category));
      options("trend", state.rows.map((r) => r.observations.trend));
      options("cover", state.rows.map((r) => r.observations.stockCoverBand), (v) => labels[v] || v);
      options("inbound", state.rows.map((r) => r.observations.inboundCompletionBand), (v) => labels[v] || v);
      $("source").textContent = `${data.source?.sourceAsOf || "-"} 기준 · 완료 ${data.source?.completedThrough || "-"}`;
      render();
    })
    .catch((e) => {
      $("status").textContent = "진단 데이터를 불러오지 못했습니다";
      console.error(e);
    });
})();
